File: Atcoder_#325/ruby/c_sensors.rb
H, W = gets.split.map(&:to_i)
grid = Array.new(H) { gets.chomp.chars }

# 移動方向を定義 (上下左右と斜め)
DIRECTIONS = [
  [-1,  0], [1,  0], [0, -1], [0,  1],  # 上下左右
  [-1, -1], [-1, 1], [1, -1], [1,  1]   # 斜め
]

def dfs(grid, x, y)
  stack = [[x, y]]
  grid[x][y] = '.'  # 訪れた場所を'.'に置き換え

  until stack.empty?
    cx, cy = stack.pop

    DIRECTIONS.each do |dx, dy|
      nx, ny = cx + dx, cy + dy
      if nx >= 0 && nx < grid.size && ny >= 0 && ny < grid[0].size && grid[nx][ny] == '#'
        grid[nx][ny] = '.'
        stack.push([nx, ny])
      end
    end
  end
end

sensor_count = 0

(0...H).each do |i|
  (0...W).each do |j|
    if grid[i][j] == '#'
      dfs(grid, i, j)
      sensor_count += 1
    end
  end
end

puts sensor_count

File: Atcoder_#325/javascript/c_sensors.js
function main(input) {
    const lines = input.trim().split('\n');
    const [H, W] = lines[0].split(' ').map(Number);
    const grid = lines.slice(1).map(line => line.split(''));

    const DIRECTIONS = [
        [-1,  0], [1,  0], [0, -1], [0,  1],  // 上下左右
        [-1, -1], [-1, 1], [1, -1], [1,  1]   // 斜め
    ];

    function dfs(grid, x, y) {
        const stack = [[x, y]];
        grid[x][y] = '.';  // 訪れた場所を'.'に置き換え

        while (stack.length > 0) {
            const [cx, cy] = stack.pop();

            DIRECTIONS.forEach(([dx, dy]) => {
                const nx = cx + dx;
                const ny = cy + dy;

                if (nx >= 0 && nx < H && ny >= 0 && ny < W && grid[nx][ny] === '#') {
                    grid[nx][ny] = '.';
                    stack.push([nx, ny]);
                }
            });
        }
    }

    let sensorCount = 0;

    for (let i = 0; i < H; i++) {
        for (let j = 0; j < W; j++) {
            if (grid[i][j] === '#') {
                dfs(grid, i, j);
                sensorCount++;
            }
        }
    }
        console.log(sensorCount);

}

// 標準入力を読み込み、main関数を呼び出し
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin', 'utf8');
main(input);
